src/DotVVM.Framework/Routing/DotvvmRouteTable.cs

```diff
@@ -93,15 +93,37 @@ public void AddGroup(string groupName, string urlPrefix, string virtualPathPrefi
         /// <param name="url">The URL.</param>
         /// <param name="virtualPath">The virtual path of the Dothtml file.</param>
         /// <param name="defaultValues">The default values.</param>
+        public void Add(string routeName, string url, string virtualPath, object defaultValues = null)
+        {
+            Add(group?.RouteNamePrefix + routeName, new DotvvmRoute(group?.UrlPrefix + url, group?.VirtualPathPrefix + virtualPath, defaultValues, GetDefaultPresenter, configuration));
+        }
+
+
+        /// <summary>
+        /// Adds the specified route name.
+        /// </summary>
+        /// <param name="routeName">Name of the route.</param>
+        /// <param name="url">The URL.</param>
+        /// <param name="defaultValues">The default values.</param>
         /// <param name="presenterFactory">The presenter factory.</param>
-        public void Add(string routeName, string url, string virtualPath, object defaultValues = null, Func<IDotvvmPresenter> presenterFactory = null)
+        public void Add(string routeName, string url, Func<IServiceProvider, IDotvvmPresenter> presenterFactory, object defaultValues = null)
         {
-            if (presenterFactory == null)
-            {
-                presenterFactory = GetDefaultPresenter;
-            }
 
-            Add(group?.RouteNamePrefix + routeName, new DotvvmRoute(group?.UrlPrefix + url, group?.VirtualPathPrefix + virtualPath, defaultValues, presenterFactory, configuration));
+            Add(group?.RouteNamePrefix + routeName, new DotvvmRoute(group?.UrlPrefix + url, group?.VirtualPathPrefix, defaultValues, presenterFactory, configuration));
+        }
+
+
+        /// <summary>
+        /// Adds the specified route name.
+        /// </summary>
+        /// <param name="routeName">Name of the route.</param>
+        /// <param name="url">The URL.</param>
+        /// <param name="presenterType">The presenter factory.</param>
+        /// <param name="defaultValues">The default values.</param>
+        public void Add(string routeName, string url, Type presenterType, object defaultValues = null)
+        {
+            Func<IServiceProvider, IDotvvmPresenter> presenterFactory = provider => (IDotvvmPresenter)provider.GetService(presenterType);
```

Comment:
I would add the following check and provide the user with a meaningful error message:

`if (!typeof(IDotvvmPresenter).IsAssignableFrom(presenterType)) throw ...`

Reply:
@tomasherceg  Please check the message text.

Reply:
I cannot see any message in the code, however I would use something like this:

"The presenter type {presenterType.ToString()} used in the route '{routeName}' does not implement IDotvvmPresenter!"